### DISCUSSION TITLE
Dynamic task mapping increases DAG File Processing time

### DISCUSSION THREAD

**Discussion by internetcoffeephone** [apache/airflow#36454] [Q&A]
SInce we started to use dynamic task mapping, we have a backfill of roughly 100k tasks that we want to re-run using DTM. Normally, this DAG parses in 1-3 seconds. There are some legacy slow top-level imports which are hard to remove.

However, when trying to run the 100k tasks, the scheduler crawls to a halt - DAG file processing time increases massively on the DAGs with lots of tasks, to the point of timeouts. We've increased the timeout to 10 minutes but even that isn't enough. We still get timeouts, and no tasks are being scheduled.

Looking at the processor, I'm not seeing things that would scale with TaskInstances (no SLA callbacks). The metadata db health is fine. DAG pickling is turned off.

I'm lacking knowledge about Airflow internals here, what possible causes are there for DAG file processing time increase so much here, and how do I remedy this?

Our Airflow version is 2.7.3

**Comment by potiuk** [comment]
Is it possible to do some more monitoring and seeing what happens while you are doing the backfill? 

I think it **could** be caused not because of the parsing time, but because of excessive database queries -  Here is one optimisation that is just about to be merged that limits the number of database queries issued during the backfill - and the number of queries is related to a number of tasks being backfilled, so this would likely match what you are observing https://github.com/apache/airflow/pull/36418#pullrequestreview-1798746446

Can you please do some tracking of the DB queries that are executed during the backfill and see if that is related ?

**Reply by internetcoffeephone** [reply]
I'd like to do that, could you explain how to do tracking of the DB queries? Does it require changing the Airflow logging_level or is it something I have to configure on the database side? Our Airflow database is postgres.

**Reply by potiuk** [reply]
It's something easiest done in your DB - you can enable server-side monitoring or logging in your DB. It can also give you some extra metrics and you can selectively enable (usually) tracking of longest running queries and the like. Generally monitoring your DB is a good idea regardless of this particular issue, so I'd strongly recommend doing it this way (and learning how to do it) as this is something that can be also used to monitor regular DB performance and issues (mis-use of indexes). I am not an expert in particular ways how to do it, but Postgres has great documentation about that, I am quite sure (and there are likely many tutorials and guidelines how to do it). Likely you can fine tune it in the way to show de-duplicated repeated queries, or only slow queries or only "suspected" queries and keep it enabled while your production operations are running.

There is also an option to enable sqlalachemy to log all queries (but this one is less selective and it will impact client side much more - likely interfering with Airflow stuff - it will drastically slow down the performance and might induce some race conditions and  possibly even deadlocks, so it should be used carefully, possibly on a staging system of yours.

This can be done with setting `echo=True` as sqlalchemy engine configuration:

https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging

And in case of Airflow it can be set via `database/sql_alchemy_engine_args` configuration parameter:

https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#sql-alchemy-engine-args

But again - this one will impact airflow processing heavily and introduce a lot of I/O contention to write to log files and extra CPU needed to format and print the log messages, so it should be used carefull on production systems.

**Reply by internetcoffeephone** [reply]
Thank you for your help!
To be precise, we're not using the `backfill` command.
Instead, we have two tasks: `get_filenames_table_name` and `process_file_table_name`. The first outputs a list of filenames, the second processes a single (big) file using dynamic task mapping.
The first task picks up several years' worth of data/filenames, because it's the first time we're running it on our existing historical data. That is what I meant by "having a backfill", apologies for any confusion.

I've done as you asked and I've observed two different cases of performance degradation.

#### Case one
Whenever I try to clear a task or mark it as success, a series of queries is ran in sequence of the form:
`SELECT airflow.task_map.length AS airflow_task_map_length FROM airflow.task_map WHERE airflow.task_map.dag_id = 'test_dag' AND airflow.task_map.run_id = 'manual__2023-12-21T16:52:38.681551+00:00' AND airflow.task_map.task_id = 'get_filenames_table_name' AND airflow.task_map.map_index < 0`

Airflow does this for _every_ task_id, so for `get_filenames_table_name` for 80 or so different table_names, even when trying to clear only one.
The queries themselves execute quickly (3ms or so), but it takes over a minute or more for the Airflow process to go through the entire sequence of tasks.

#### Case two
Whenever the scheduler is supposed to schedule a task, the following slow (8+ seconds) query runs over and over:
`SELECT airflow.task_instance.try_number, airflow.task_instance.task_id, airflow.task_instance.dag_id, airflow.task_instance.run_id, airflow.task_instance.map_index, airflow.task_instance.start_date, airflow.task_instance.end_date,
airflow.task_instance.duration, airflow.task_instance.state, airflow.task_instance.max_tries, airflow.task_instance.hostname, airflow.task_instance.unixname, airflow.task_instance.job_id, airflow.task_instance.pool,
airflow.task_instance.pool_slots, airflow.task_instance.queue, airflow.task_instance.priority_weight, airflow.task_instance.operator, airflow.task_instance.custom_operator_name, airflow.task_instance.queued_dttm,
airflow.task_instance.queued_by_job_id, airflow.task_instance.pid, airflow.task_instance.executor_config, airflow.task_instance.updated_at, airflow.task_instance.external_executor_id, airflow.task_instance.trigger_id,
airflow.task_instance.trigger_timeout, airflow.task_instance.next_method, airflow.task_instance.next_kwargs, dag_run_1.state AS state_1, dag_run_1.id, dag_run_1.dag_id AS dag_id_1, dag_run_1.queued_at, dag_run_1.execution_date, dag_run_1.start_date
AS start_date_1, dag_run_1.end_date AS end_date_1, dag_run_1.run_id AS run_id_1, dag_run_1.creating_job_id, dag_run_1.external_trigger, dag_run_1.run_type, dag_run_1.conf, dag_run_1.data_interval_start, dag_run_1.data_interval_end,
dag_run_1.last_scheduling_decision, dag_run_1.dag_hash, dag_run_1.log_template_id, dag_run_1.updated_at AS updated_at_1 FROM airflow.task_instance JOIN airflow.dag_run AS dag_run_1 ON dag_run_1.dag_id = airflow.task_instance.dag_id AND
dag_run_1.run_id = airflow.task_instance.run_id WHERE airflow.task_instance.dag_id = 'test_dag' AND airflow.task_instance.run_id = 'manual__2023-12-21T16:52:38.681551+00:00' AND (airflow.task_instance.state IN ('removed', 'scheduled',
'queued', 'running', 'success', 'restarting', 'failed', 'up_for_retry', 'up_for_reschedule', 'upstream_failed', 'skipped', 'deferred', 'shutdown') OR airflow.task_instance.state IS NULL)`

It seems obvious that such a join becomes very slow if the `airflow.task_instance` table becomes very large for a given `airflow.task_instance.run_id`.

Another minor bug I found: marking the task group that contains both `get_filenames_table_name` and `process_file_table_name` as success only marks the `get_filenames_table_name` task as success.

**Reply by internetcoffeephone** [reply]
A small addition to case two, the query:
`SELECT airflow.dag_run.id FROM airflow.dag_run WHERE airflow.dag_run.dag_id = 'test_dag' AND airflow.dag_run.execution_date >= '2023-12-21T16:52:38.681551+00:00'::timestamptz AND airflow.dag_run.execution_date <='2023-12-21T16:52:38.681551+00:00'::timestamptz FOR UPDATE`
is also ran and hangs for 15-20+ minutes. The FOR UPDATE is never released with a COMMIT, which causes hangs in other parts of Airflow. My guess is that the calling process has probably been killed by that point due to a timeout. These queries do eventually disappear though.

Our `AIRFLOW__WEBSERVER__WEB_SERVER_MASTER_TIMEOUT`, `AIRFLOW__WEBSERVER__WEB_SERVER_WORKER_TIMEOUT`, `AIRFLOW__CORE__DAG_FILE_PROCESSOR_TIMEOUT` and `AIRFLOW__CORE__DAGBAG_IMPORT_TIMEOUT` are set to 600 to debug this issue, which could explain why it hangs for 10 minutes. Possibly the other 5-10 are explained by Postgres timeout settings? Either way, this hang is probably caused by the queries written in my previous reply.

**Reply by internetcoffeephone** [reply]
I collected 23 minutes' worth of logs where Airflow is attempting to scheduling tens of thousands of tasks (of which +-20 can run concurrently), where I am also clearing failed tasks. Then, I ran the logs through pgbadger.
[pgbader_out.zip](https://github.com/apache/airflow/files/13863515/pgbader_out.zip)
Two things stand out:

1. There are no truly slow queries.
![Screenshot 2024-01-05 at 16 17 30](https://github.com/apache/airflow/assets/5096835/fe2e52a8-715d-4672-b404-3c7fcb382ad6)

2. The most-called query is called over 10x more often than the next-most called query.
![Screenshot 2024-01-05 at 16 12 56](https://github.com/apache/airflow/assets/5096835/7b3d9abf-52f7-421e-8ad0-887e2ffb72e6)

**Reply by internetcoffeephone** [reply]
I used py-spy to profile the scheduler process and create a flame graph. This confirms the above data: the scheduler is calling the function [_get_length()](https://github.com/apache/airflow/blob/b75f9e880614fa0427e7d24a1817955f5de658b3/airflow/models/expandinput.py#L131) excessively frequently. @uranusjr it seems like you foresaw this problem in your comment [here](https://github.com/apache/airflow/blob/b75f9e880614fa0427e7d24a1817955f5de658b3/airflow/models/expandinput.py#L138).

![profile](https://github.com/apache/airflow/assets/5096835/c2b31cb3-6b2a-4c0a-b51e-be14448e92d4)

**Reply by potiuk** [reply]
cc: @uranusjr -> any thoughts on that one?

**Comment by potiuk** [comment]
I also created https://github.com/apache/airflow/pull/36483 targeted to be released in 2.8.1 as part of our documentation, largely explaining the same  I just did above in our documentation.
